buffer/src/test/java/io/netty/buffer/AlignedPooledByteBufAllocatorTest.java

```diff
@@ -63,4 +66,58 @@ public void testCorrectElementSize() {
         assertTrue(a.release());
         assertTrue(b.release());
     }
+    @Test
+    public void testDirectSubpageReleaseLock() throws Exception {
+        int initialCapacity = 0;
+        int directMemoryCacheAlignment = 32;
+        PooledByteBufAllocator allocator = new PooledByteBufAllocator(
+                true,
+                0,
+                1,
+                PooledByteBufAllocator.defaultPageSize(),
+                PooledByteBufAllocator.defaultMaxOrder(),
+                0,
+                0,
+                false,
+                directMemoryCacheAlignment);
+
+        final PooledByteBuf<?> byteBuf = pooledByteBuf(allocator.directBuffer(initialCapacity, 16));
+        // Get the smallSubpagePools[] array in arena.
+        @SuppressWarnings("unchecked")
+        PoolSubpage<byte[]>[] smallSubpagePools = (PoolSubpage<byte[]>[]) byteBuf.chunk.arena.smallSubpagePools;
+        PoolSubpage<byte[]> head = null;
+        for (PoolSubpage<byte[]> subpage : smallSubpagePools) {
+            if (subpage.next != subpage) {
+                // Find the head subpage which the byteBuf belongs to.
+                head = subpage;
+                break;
+            }
+        }
+        assertNotNull(head);
+        Thread t1 = new Thread(new Runnable() {
+            @Override
+            public void run() {
+                // Because the head subpage was already locked in the main thread, so this should hang and wait.
+                byteBuf.release();
+            }
+        });
+        t1.setDaemon(true);
+        // Intentionally lock the head subpage in main thread.
+        head.lock();
+        try {
+            t1.start();
+            // Wait max 3 seconds.
+            t1.join(1000 * 3);
```

Comment:
Instead of having a mandatory 3-second wait here, we could make `PoolSubpage.lock` package-private and spin on a `while (!head.lock.hasQueuedThread(t1)) { ... }` condition, until some deadline.

Reply:
But imho, we still need to specified a mandatory wait 'time' or 'times' as the deadline signal, to decide when to end the `while (!head.lock.hasQueuedThread(t1)) { ... }` looping.

Reply:
But imho, even though, we still need a mandatory wait 'time' or 'times' to decide when to end the `while (!head.lock.hasQueuedThread(t1)) { ... }` looping.

Reply:
The code can set itself a deadline and check it with `System.nanoTime()`. I think that should suffice to avoid infinite looping.

Reply:
@chrisvest , done, PTAL.